ui/src/app/machines/views/MachineDetails/MachineNetwork/NetworkTable/NetworkTable.tsx

```diff
@@ -58,184 +66,240 @@ type SortKey = keyof NetworkRowSortData;
-                    className="p-link--muted"
-                    route={`/vlan/${vlan.id}`}
+              data-test="dhcp"
+              icon={
+                vlan && vlan.relay_vlan ? (
+                  <Tooltip
+                    position="btm-right"
+                    message={getDHCPStatus(vlan, vlans, fabrics, true)}
                   >
-                    {getVLANDisplay(vlan)}
-                  </LegacyLink>
+                    <Icon name="information" />
+                  </Tooltip>
                 ) : null
               }
+              iconSpace={true}
+              primary={getDHCPStatus(vlan, vlans, fabrics)}
             />
-          ),
-        },
-        {
-          content: !isABondOrBridgeParent && (
-            <SubnetColumn nic={nic} systemId={machine.system_id} />
-          ),
-        },
-        {
-          content: !isABondOrBridgeParent && (
-            <IPColumn nic={nic} systemId={machine.system_id} />
-          ),
-        },
-        {
-          content:
-            !isABondOrBridgeParent && fabricsLoaded && vlansLoaded ? (
-              <DoubleRow
-                data-test="dhcp"
-                icon={
-                  vlan && vlan.relay_vlan ? (
-                    <Tooltip
-                      position="btm-right"
-                      message={getDHCPStatus(vlan, vlans, fabrics, true)}
-                    >
-                      <Icon name="information" />
-                    </Tooltip>
-                  ) : null
-                }
-                iconSpace={true}
-                primary={getDHCPStatus(vlan, vlans, fabrics)}
-              />
-            ) : null,
-        },
-        {
-          className: "u-align--right",
-          content: !isABondOrBridgeParent && (
-            <NetworkTableActions nic={nic} systemId={machine.system_id} />
-          ),
-        },
-      ],
-      key: nic.id,
-      sortData: {
-        name: nic.name,
-        pxe: isBoot,
-        speed: nic.link_speed,
-        type: null,
-        fabric: null,
-        subnet: null,
-        ip: null,
-        dhcp: null,
+          ) : null,
+      },
+      {
+        className: "u-align--right",
+        content: !isABondOrBridgeParent && (
+          <NetworkTableActions nic={nic} systemId={machine.system_id} />
+        ),
       },
+    ],
+    key: nic.name,
+    sortData: {
+      name: nic.name,
+      pxe: isBoot,
+      speed: nic.link_speed,
+      type: null,
+      fabric: null,
+      subnet: null,
+      ip: null,
+      dhcp: null,
+    },
+  };
+};
+
+const generateRows = (
+  machine: Machine,
+  fabrics: Fabric[],
+  vlans: VLAN[],
+  fabricsLoaded: boolean,
+  vlansLoaded: boolean
+): NetworkRow[] => {
+  if (!machine || !("interfaces" in machine)) {
+    return [];
+  }
+  const interfaces: NetworkLinkInterface[] = [];
+  // Create a list of interfaces and aliases to use to generate the table rows.
+  machine.interfaces.forEach((nic: NetworkInterface) => {
```

Comment:
It seems like there's a bit too much object manipulation happening in this function, so it's hard to know exactly what is going where. Introducing a new type `NetworkLinkInterface` mitigates this a bit, however it introduces its own weirdness because it feels like the type is masquerading as something defined in the API but it really only exists as a convenience for generating table rows in the UI.

I ran into a similar issue in the storage tab where disks (interfaces) can have partitions (links) but the hierarchy isn't really visualised in the table. Originally I wrote it similar to this where a common "row" object is defined (like `NetworkLinkInterface`, but more bespoke), but when I started implementing the actions it became apparent that they're two separate entities and should be treated as such. My solution was to keep the data from the API intact as long as possible by rewriting most of the row generation logic so that the functions accepted both a `Disk` and `Partition` type, and only at the render level make the differences apparent.

I think it works better and is simpler. With the intermediary type, the data flow went like:
- Get disk or partition data from API
- Convert disk or partition into intermediary type
- Determine whether the intermediary type was originally a disk or partition
- Render the disk or partition data

And without the intermediary type it goes like:
- Get disk or partition data from API
- Determine whether it's a disk or partition
- Render the disk or partition data

This is all to say I'm not actually sure if it's the best way to address this, because now the conditionals have just been moved from the row object generation to the render function so there's a bunch of conditionals like `{isDisk(storageDevice) ? diskValue : partitionValue}` which is its own version of weirdness. But I think the most important thing is that the data from the API is changed as little as possible.

I'm interested to hear what you think, and sorry for the wall of text!

Reply:
Thanks @Caleb-Ellis I think it's good to try and get this right so that it's not as hard to untangle as the legacy version of this code (where you never really know where the row data has come from).

I'll have another try at using the original types and see how I go.